src/Features/Core/Portable/FindUsages/SourceReferenceItem.cs

```diff
@@ -25,11 +26,48 @@ internal sealed class SourceReferenceItem
         /// </summary>
         public bool IsWrittenTo { get; }
 
+        /// <summary>
+        /// Additional information about the reference.
+        /// Each entry represents a key-values pair of data. For example, consider the below entry:
+        ///     { "ValueUsageInfo" } = { "Read", "Write" }
+        /// This entry indicates that the reference has additional value usage information which indicate
+        /// it is a read/write reference, such as say 'a++'.
+        /// </summary>
+        public ImmutableDictionary<string, ImmutableArray<string>> ReferenceInfo { get; }
+
+        [Obsolete]
         public SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, bool isWrittenTo)
         {
             Definition = definition;
             SourceSpan = sourceSpan;
             IsWrittenTo = isWrittenTo;
+            ReferenceInfo = ImmutableDictionary<string, ImmutableArray<string>>.Empty;
+        }
+
+        public SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, ImmutableDictionary<string, ImmutableArray<string>> referenceInfo)
+        {
+            Definition = definition;
+            SourceSpan = sourceSpan;
+            ReferenceInfo = referenceInfo ?? throw new ArgumentNullException(nameof(referenceInfo));
+        }
+
+        internal SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, ValueUsageInfo valueUsageInfo)
```

Comment:
I have maked the the ctor with IsWrittenTo as obsolete. @CyrusNajmabadi told me that TS uses it, so I did not remove it. Hopefully they can move to new ctor and then we can completely remove IsWrittenTo and obsolete ctor.

Reply:
Note: i'm not certain if they use it.  might as well check.  if tehy don't, def just remove :)

Reply:
Verified that TypeScript does still use the constructor `public SourceReferenceItem(DefinitionItem definition, DocumentSpan sourceSpan, bool isWrittenTo)`. I am going to retain the existing approach.

---
In reply to: [220759901](https://github.com/dotnet/roslyn/pull/30155#discussion_r220759901) [](ancestors = 220759901)

Reply:
on the fence about ValueUsageInfo being baked into this API directly.  I think i'd prefer if it just took a `MultiDictionary<string, string> ReferenceInfo` as an argument instead.

Other minor note: Is MultiDictionary mutable?  If so... i may have to retract my position about using it since we'd want immutable data here.  If it is immutbale, great!

Reply:
> Is MultiDictionary mutable

Yes, it is. I will revert it back to `ImmutableDictionary<string, ImmutableArray<string>>`.

Reply:
Darn... sorry about that :(

Reply:
We invoke this from multiple places and we are caching the mapping from ValueUsageInfo to the reference info values map here, so for now I will retain this here.

---
In reply to: [222055535](https://github.com/dotnet/roslyn/pull/30155#discussion_r222055535) [](ancestors = 222055535)

Reply:
No issues :) moved back to ImmutableDictionary.

---
In reply to: [222063382](https://github.com/dotnet/roslyn/pull/30155#discussion_r222063382) [](ancestors = 222063382)